File: red/src/managed-access.ts
import {constants,openSync,closeSync,writeSync,fsyncSync,renameSync,unlinkSync,lstatSync,mkdirSync} from 'node:fs';
import {dirname,resolve} from 'node:path';
import {randomUUID} from 'node:crypto';
import {registry} from './index.ts';
import {parseStateEnvelope} from './backend.ts';
type Map=Record<string,any>;
const obj=(v:any):v is Map=>v!==null&&typeof v==='object'&&!Array.isArray(v);
const safe=(v:any)=>typeof v==='string'&&/^[A-Za-z0-9][A-Za-z0-9_-]{0,62}$/.exec(v)?.[0]===v;
const demand=(v:any,message='invalid managed outputs')=>{if(!v)throw Error(message);};
export function managed_kubeconfig_path(opts:Map):string {demand(typeof opts.workdir==='string'&&!!opts.workdir.trim()&&safe(opts.profile),'invalid managed access path');return resolve(opts.workdir,opts.profile,'kubeconfig');}
export function publicParams(value:any,provider:string):Map {
 const required=['provider','kind','name','cluster_id','endpoint'];demand(obj(value)&&required.every(k=>Object.hasOwn(value,k))&&Object.keys(value).every(k=>[...required,'pod_cidr','service_cidr'].includes(k)));
 demand(value.provider===provider&&value.kind==='managed-kubernetes'&&safe(value.name));demand(typeof value.cluster_id==='string'&&/^[A-Za-z0-9][A-Za-z0-9_.:-]{0,255}$/.exec(value.cluster_id)?.[0]===value.cluster_id);
 demand(typeof value.endpoint==='string');const url=new URL(value.endpoint);demand(url.protocol==='https:'&&url.hostname&&!url.username&&!url.password&&!url.hash&&!url.search);
 for(const key of ['pod_cidr','service_cidr'])if(Object.hasOwn(value,key)){const v=value[key];demand(typeof v==='string');const m=/^(\d+)\.(\d+)\.(\d+)\.(\d+)\/(\d+)$/.exec(v);demand(m&&m[0]===v);const a=m!.slice(1).map(Number);demand(a.slice(0,4).every(n=>n>=0&&n<=255)&&a[4]!<=32&&a.map(String).join('.')===v.replace('/','.'));const n=((a[0]!<<24)|(a[1]!<<16)|(a[2]!<<8)|a[3]!)>>>0;demand((a[4]===0?n===0:(n&((2**(32-a[4]!)-1)>>>0))===0));}
 return {...value};
}
export class AccessDecoder {
 private content:Uint8Array|null=null;
 private secrets:string[];
 constructor(private opts:Map,environment:Record<string,string|undefined>=process.env){const keys=[...(registry.compute as Map)[opts["provider-compute"]].secrets,...(opts["provider-backend"]==="r2"?["r2-access-key-id","r2-secret-access-key"]:[])];this.secrets=keys.map(k=>environment["COLORS_PAR_"+k.toUpperCase().replaceAll("-","_")]).filter((v):v is string=>typeof v==="string"&&!!v);}
 decode=(text:string):Map=>{
  const {document,params}=parseStateEnvelope(text);const publicValue=publicParams(params,this.opts['provider-compute']),outputs=document.outputs;demand(Object.keys(outputs).length===2&&Object.hasOwn(outputs,'params')&&Object.hasOwn(outputs,'kubeconfig_b64')&&(outputs.params.sensitive??false)===false);const entry=outputs.kubeconfig_b64;demand(obj(entry)&&entry.sensitive===true&&typeof entry.value==='string'&&entry.value.length<=2796204);
  const raw=Buffer.from(entry.value,'base64');demand(raw.toString('base64')===entry.value&&raw.length>0&&raw.length<=2097152);const content=new TextDecoder('utf-8',{fatal:true,ignoreBOM:true}).decode(raw);demand(!content.includes('\0')&&!content.startsWith('\ufeff'));
  demand(!this.secrets.some(secret=>content.includes(secret)||content.includes(JSON.stringify(secret).slice(1,-1))));const config=Bun.YAML.parse(content) as Map;demand(obj(config)&&config.kind==='Config'&&config.apiVersion==='v1');demand(Array.isArray(config.clusters)&&config.clusters.length===1&&obj(config.clusters[0])&&obj(config.clusters[0].cluster));const cluster=config.clusters[0].cluster;demand(typeof cluster.server==='string'&&cluster.server.replace(/\/+$/,'')===params.endpoint.replace(/\/+$/,'')&&(cluster['insecure-skip-tls-verify']??false)===false);const contexts=config.contexts,users=config.users;demand(Array.isArray(contexts)&&contexts.length===1&&Array.isArray(users)&&users.length===1);const context=contexts[0],user=users[0];demand(obj(context)&&obj(user)&&obj(context.context)&&obj(user.user));const names=[config.clusters[0].name,context.name,user.name];demand(names.every(name=>typeof name==='string'&&!!name.trim()));demand(config['current-context']===names[1]&&context.context.cluster===names[0]&&context.context.user===names[2]);const credentials=user.user;demand((typeof credentials.token==='string'&&!!credentials.token.trim())||['client-certificate-data','client-key-data'].every(key=>typeof credentials[key]==='string'&&!!credentials[key].trim()));const forbidden=new Set(['exec','auth-provider','tokenFile','client-certificate','client-key','certificate-authority','proxy-url','tls-server-name']),seen=new Set<object>(),pending:any[]=[config];
  while(pending.length){const v=pending.pop();if(v&&typeof v==='object'){demand(!seen.has(v)&&seen.size<10000);seen.add(v);if(!Array.isArray(v))demand(Object.keys(v).every(k=>!forbidden.has(k)));pending.push(...Object.values(v));}}
  this.content=raw;return {params:publicValue};
 };
 write():string {
  demand(this.content!==null,'managed access unavailable');const target=managed_kubeconfig_path(this.opts);let part=target;
  for(;;){try{const stat=lstatSync(part);demand(!stat.isSymbolicLink(),'invalid managed access path');if(part===target)demand(stat.isFile()&&stat.uid===process.getuid!(),'invalid managed access path');}catch(error){if((error as any).code!=='ENOENT')throw error;}const parent=dirname(part);if(parent===part)break;part=parent;}
  mkdirSync(dirname(target),{mode:0o700,recursive:true});const temporary=resolve(dirname(target),`.kubeconfig-${randomUUID()}`);let fd:number|undefined;
  try{fd=openSync(temporary,constants.O_WRONLY|constants.O_CREAT|constants.O_EXCL|constants.O_NOFOLLOW,0o600);let offset=0;while(offset<this.content!.length){const count=writeSync(fd,this.content!,offset,this.content!.length-offset);demand(count>0);offset+=count;}fsyncSync(fd);closeSync(fd);fd=undefined;renameSync(temporary,target);return target;}finally{if(fd!==undefined)closeSync(fd);try{unlinkSync(temporary);}catch{}this.content=null;}
 }
}

File: blue/src/colors_compute/managed_access.py
"""Validated managed-cluster outputs and a fixed private kubeconfig sink."""
import base64
import binascii
import ipaddress
import json
import os
from pathlib import Path
import re
import tempfile
import yaml
from urllib.parse import urlsplit

from .backend import _params
from .contract import _safe


def managed_kubeconfig_path(opts):
    workdir, profile = opts.get('workdir'), opts.get('profile')
    if not isinstance(workdir, str) or not workdir.strip() or not _safe(profile):
        raise ValueError('invalid managed access path')
    return str(Path(workdir).absolute() / profile / 'kubeconfig')


def public_params(value, provider):
    required = {'provider', 'kind', 'name', 'cluster_id', 'endpoint'}
    if not isinstance(value, dict) or not required <= set(value) or set(value) - required - {'pod_cidr', 'service_cidr'}:
        raise ValueError('invalid managed outputs')
    if value['provider'] != provider or value['kind'] != 'managed-kubernetes' or not _safe(value['name']):
        raise ValueError('invalid managed outputs')
    if not isinstance(value['cluster_id'], str) or not re.fullmatch(r'[A-Za-z0-9][A-Za-z0-9_.:-]{0,255}', value['cluster_id']):
        raise ValueError('invalid managed outputs')
    endpoint = value['endpoint']
    if not isinstance(endpoint, str):
        raise ValueError('invalid managed outputs')
    url = urlsplit(endpoint)
    if url.scheme != 'https' or not url.hostname or url.username or url.password or url.fragment or url.query:
        raise ValueError('invalid managed outputs')
    for key in ('pod_cidr', 'service_cidr'):
        if key in value and (not isinstance(value[key], str) or ipaddress.ip_network(value[key], strict=True).version != 4 or str(ipaddress.ip_network(value[key], strict=True)) != value[key]):
            raise ValueError('invalid managed outputs')
    return dict(value)


class AccessDecoder:
    def __init__(self, opts, environment=None):
        self.opts = opts
        env = os.environ if environment is None else environment
        from .contract import registry
        keys = list(registry()["compute"][opts["provider-compute"]]["tofu-env"])
        self._secrets = [env.get("COLORS_PAR_" + key.upper().replace("-", "_")) for key in keys + list(registry()["backend"].get(opts.get("provider-backend"), {}).get("secrets", []))]
        self._secrets = [value for value in self._secrets if isinstance(value, str) and value]
        self._content = None

    def __call__(self, text):
        params = public_params(_params(text), self.opts['provider-compute'])
        outputs = json.loads(text)['outputs']
        if set(outputs) != {'params', 'kubeconfig_b64'}:
            raise ValueError('invalid managed outputs')
        if outputs['params'].get('sensitive', False) is not False:
            raise ValueError('invalid managed outputs')
        entry = outputs['kubeconfig_b64']
        if not isinstance(entry, dict) or entry.get('sensitive') is not True or not isinstance(entry.get('value'), str) or len(entry['value']) > 2796204:
            raise ValueError('invalid managed outputs')
        try:
            raw = base64.b64decode(entry['value'], validate=True)
            content = raw.decode('utf-8')
        except (ValueError, UnicodeError, binascii.Error):
            raise ValueError('invalid managed outputs') from None
        if base64.b64encode(raw).decode('ascii') != entry['value'] or not raw or len(raw) > 2097152 or '\x00' in content or content.startswith('\ufeff'):
            raise ValueError('invalid managed outputs')
        if any(secret in content or json.dumps(secret)[1:-1] in content for secret in self._secrets):
            raise ValueError('invalid managed outputs')
        # Parse both YAML and JSON forms; flow mappings cannot bypass this guard.
        config = yaml.safe_load(content)
        if not isinstance(config, dict) or config.get('kind') != 'Config' or config.get('apiVersion') != 'v1':
            raise ValueError('invalid managed outputs')
        clusters = config.get('clusters')
        if not isinstance(clusters, list) or len(clusters) != 1 or not isinstance(clusters[0], dict):
            raise ValueError('invalid managed outputs')
        cluster = clusters[0].get('cluster')
        if not isinstance(cluster, dict) or cluster.get('server', '').rstrip('/') != params['endpoint'].rstrip('/') or cluster.get('insecure-skip-tls-verify', False) is not False:
            raise ValueError('invalid managed outputs')
        contexts, users = config.get('contexts'), config.get('users')
        if not isinstance(contexts, list) or len(contexts) != 1 or not isinstance(users, list) or len(users) != 1:
            raise ValueError('invalid managed outputs')
        context, user = contexts[0], users[0]
        if not isinstance(context, dict) or not isinstance(user, dict) or not isinstance(context.get('context'), dict) or not isinstance(user.get('user'), dict):
            raise ValueError('invalid managed outputs')
        names = [clusters[0].get('name'), context.get('name'), user.get('name')]
        if any(not isinstance(name, str) or not name.strip() for name in names):
            raise ValueError('invalid managed outputs')
        if config.get('current-context') != names[1] or context['context'].get('cluster') != names[0] or context['context'].get('user') != names[2]:
            raise ValueError('invalid managed outputs')
        credentials = user['user']
        token = isinstance(credentials.get('token'), str) and bool(credentials['token'].strip())
        certificate = all(isinstance(credentials.get(key), str) and bool(credentials[key].strip()) for key in ('client-certificate-data', 'client-key-data'))
        if not (token or certificate):
            raise ValueError('invalid managed outputs')
        forbidden = {'exec', 'auth-provider', 'tokenFile', 'client-certificate', 'client-key', 'certificate-authority', 'proxy-url', 'tls-server-name'}
        seen, pending = set(), [config]
        while pending:
            value = pending.pop()
            if isinstance(value, (dict, list)):
                if id(value) in seen:
                    raise ValueError('invalid managed outputs')
                seen.add(id(value))
                if len(seen) > 10000:
                    raise ValueError('invalid managed outputs')
                if isinstance(value, dict):
                    if any(not isinstance(key, str) or key in forbidden for key in value):
                        raise ValueError('invalid managed outputs')
                    pending.extend(value.values())
                else:
                    pending.extend(value)
        self._content = raw
        return {'params': params}

    def write(self):
        if self._content is None:
            raise ValueError('managed access unavailable')
        target = Path(managed_kubeconfig_path(self.opts))
        # Reject every symlink ancestor, including a preexisting target.
        for part in [*reversed(target.parents), target]:
            if part.is_symlink():
                raise ValueError('invalid managed access path')
        target.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
        if target.exists() and (not target.is_file() or target.stat().st_uid != os.getuid()):
            raise ValueError('invalid managed access path')
        fd, temporary = tempfile.mkstemp(prefix='.kubeconfig-', dir=target.parent)
        try:
            with os.fdopen(fd, 'wb') as stream:
                stream.write(self._content)
                stream.flush()
                os.fsync(stream.fileno())
            os.replace(temporary, target)
        finally:
            if os.path.exists(temporary):
                os.unlink(temporary)
            self._content = None
        return str(target)
